elements/dash-tree/dash-tree.js

```diff
@@ -0,0 +1,436 @@
+      this.setAttribute('aria-activedescendant', treeItem.id);
+
+      // There can be only one active item at a time.
+      // Find any previous active item and remove its active class.
+      const activeItem = this.querySelector('.active');
+      if (activeItem)
+        activeItem.classList.remove('active');
+
+      treeItem.classList.add('active');
+    }
+
+    /**
+     * Flip the `DashTreeItem` between open and closed states.
+     */
+    _toggleTreeItem() {
+      const treeItem = this._currentTreeItem();
+      if (isExpanded(treeItem)) {
+        this._collapseTreeItem();
+      } else {
+        this._expandTreeItem();
+      }
+    }
+
+    /**
+     * If focus is on an open node, closes the node.
+     * If focus is on a child node that is also either an end node or a closed
+     * node, move focus to its parent node.
+     */
+    _collapseTreeItem() {
+      const treeItem = this._currentTreeItem();
+      if (treeItem.firstElementChild && isExpanded(treeItem)) {
+        treeItem.setAttribute('aria-expanded', 'false');
+        return;
+      }
+      // Walk up the tree till you find the parent `<dash-tree-item>`.
+      // If this is a root node, do nothing. Otherwise, collapse the parent
+      // and move focus to it.
+      let parent = treeItem.parentElement;
+      if (parent === this)
+        return;
+      while (!isTreeItem(parent)) {
+        parent = parent.parentElement;
+      }
+      parent.setAttribute('aria-expanded', 'false');
+      this._focusTreeItem(parent);
+    }
+
+    /**
+     * If focus is on a closed node, opens the node.
+     */
+    _expandTreeItem() {
+      const treeItem = this._currentTreeItem();
+      if (treeItem.firstElementChild) {
+        treeItem.setAttribute('aria-expanded', 'true');
+      }
+    }
+
+    /**
+     * Perform the default action for a treeitem. If the item is a parent
+     * node, toggle its expanded/collapsed state. If the item is an end
+     * node, dispatch an event with a reference to the node. If this was
+     * a file picker, an application could listen for this event and open
+     * the file based on the item's name.
+     */
+    _selectTreeItem() {
+      // There can only be one selected element at time.
+      // Look at all the children and remove `aria-selected` and the `.selected`
+      // class from any element that has them.
+      this.querySelectorAll('[aria-selected], .selected')
+        .forEach(item => {
+          item.removeAttribute('aria-selected');
+          item.classList.remove('selected');
+        });
+
+      const treeItem = this._currentTreeItem();
+      treeItem.setAttribute('aria-selected', 'true');
+      treeItem.classList.add('selected');
+
+      // Dispatch a non-bubbling event containing a reference to the selected
+      // node. The reason to choose non-bubbling is explained in this post
+      // https://medium.com/dev-channel/custom-elements-that-work-anywhere-898e1dd2bc48#.w6ww4mgfc
```

Comment:
maybe rater `[this post](... link ...)`? Looks better on [the page](http://dash-elements.surma.link/a6f34742f35b7aba2a838ccbea368c29964c8f9c/dash-tree.html).